bytestring/jvm/src/ByteStringJvmExt.kt

```diff
@@ -20,3 +24,103 @@ public fun ByteString.decodeToString(charset: Charset): String = getBackingArray
+ */
+@OptIn(UnsafeByteStringApi::class)
+public fun ByteString.asReadOnlyByteBuffer(): ByteBuffer {
+    val data: ByteArray
+
+    UnsafeByteStringOperations.withByteArrayUnsafe(this) {
+        data = it
+    }
+
+    return ByteBuffer.wrap(data).asReadOnlyBuffer()
+}
+
+/**
+ * Reads [length] bytes of data from [this] ByteBuffer starting from the current position and
+ * wraps them into a new [ByteString].
+ *
+ * Upon successful execution, current position will advance by [length].
+ *
+ * @throws IndexOutOfBoundsException when [length] has negative value or its value exceeds [ByteBuffer.remaining]
+ */
+@OptIn(UnsafeByteStringApi::class)
+public fun ByteBuffer.getByteString(length: Int = remaining()): ByteString {
+    if (length < 0) {
+        throw IndexOutOfBoundsException("length should be non-negative (was $length)")
+    }
+    if (remaining() < length) {
+        throw IndexOutOfBoundsException("length ($length) exceeds remaining bytes count ({${remaining()}})")
+    }
+    val bytes = ByteArray(length)
+    get(bytes)
+    return UnsafeByteStringOperations.wrapUnsafe(bytes)
+}
+
+/**
+ * Reads [length] bytes of data from [this] ByteBuffer starting from [at] index and
+ * wraps them into a new [ByteString].
+ *
+ * This function does not update [ByteBuffer.position].
+ *
+ * @throws IndexOutOfBoundsException when [at] is negative, greater or equal to [ByteBuffer.limit]
+ * or [at] + [length] exceeds [ByteBuffer.limit].
+ */
+@OptIn(UnsafeByteStringApi::class)
+public fun ByteBuffer.getByteString(at: Int, length: Int): ByteString {
+    checkIndexAndCapacity(at, length)
+    val bytes = ByteArray(length)
+    // Absolute get(byte[]) was added only in JDK 13
+    for (i in 0..<length) {
+        bytes[i] = get(at + i)
+    }
+    return UnsafeByteStringOperations.wrapUnsafe(bytes)
+}
+
+/**
+ * Writes [string] into [this] ByteBuffer starting from the current position.
+ *
+ * Upon successfully execution [ByteBuffer.position] will advance by the length of [string].
+ *
+ * @throws java.nio.ReadOnlyBufferException when [this] buffer is read-only
+ * @throws java.nio.BufferOverflowException when [string] can't fit into remaining space of this buffer
+ */
+@OptIn(UnsafeByteStringApi::class)
+public fun ByteBuffer.putByteString(string: ByteString) {
+    UnsafeByteStringOperations.withByteArrayUnsafe(string) {
+        put(it)
+    }
+}
+
+/**
+ * Writes [string] into [this] ByteBuffer starting from position [at].
+ *
+ * This function does not update [ByteBuffer.position].
+ *
+ * @throws java.nio.ReadOnlyBufferException when [this] buffer is read-only
+ * @throws IndexOutOfBoundsException when [at] is negative, exceeds [ByteBuffer.limit], or
+ * [at] + [ByteString.size] exceeds [ByteBuffer.limit]
+ */
+public fun ByteBuffer.putByteString(at: Int, string: ByteString) {
+    checkIndexAndCapacity(at, string.size)
+    // Absolute get(byte[]) was added only in JDK 13
```

Comment:
```suggestion
    // Absolute put(byte[]) was added only in JDK 16
```
https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/nio/ByteBuffer.html#put(int,java.nio.ByteBuffer,int,int)